### ISSUE TITLE
Custom ACL

### ISSUE THREAD

**Issue by Sepremex** [alexusmai/laravel-file-manager#195]
It is possible that I'm looking this in the wrong way...
When user login I get all their information, I have a multilevel access.
Admin -- access to all
Manager -- access to few based on "project"
Crew -- access based on "manager/project"
User -- access to it's own
-----
then I have folders, [projects, users]
if a manager is uploading images, videos or files, this manager can only have access to one particular project which have access to one folder at that moment...
A manager can only have 1 project and never more than that
if the manager is remove from a project, access to assets from that project wont be available BUT has access to personal files, the same apply or the crew under manager user.

I have no idea how to configure this type of access... 
I'm sure this is not on the configuration file but in the Controller for project and controller for manager/crew ...
I haven't done nothing yet I'm just thinking about it while building my database and other files for this project.
I have used this library on other project with basic configuration but nothing like this...

Can someone point me in the right direction please.

**Comment by BRAiNCHiLD95** [comment]
I'm in the same boat as you - I'm going to try and implement this using Laravel 7. If I manage to successfully get this working - I'll share my approach! :)
But, if anyone has any insights on how to do this - I'd be super grateful!!

**Comment by BRAiNCHiLD95** [comment]
Hey @Sepremex - I've managed to implement what I needed. 

## ACL
This was quite annoying to do, but figured it out after fixing the "leftPath" (check middleware section below) to **EXACTLY match** the path in ACL rules (I had an extra "/" - and it kept failing).

In app/Http, I created a custom ACL Repository based on the docs provided by the maintainer.

In my case (like you), I wanted to use user->role instead of user->id - so I modified getUserID() function as follows -

`public function getUserID()    {`
&nbsp;&nbsp;&nbsp;&nbsp;`return Auth::user()->role;`
`}`

This function is used mainly to cache/index the rules - so make sure they're properly set-up.
Next, I rewrote the main getRules() function -

using **if** conditions to return the right rules array (I have super-users, admins, staff and users - and I have declared constants for ROLE_SUPERVISOR, ROLE_ADMIN, and ROLE_STAFF in my User model).

`$roomPath = // logic to get the path (in my case - it's a regEx based matching on the HTTP_REFERRER`

`if (Auth::user()->role === User::ROLE_SUPERVISOR)    {`
` // complete access`
&nbsp;&nbsp;&nbsp;&nbsp;`return [`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`['disk' => 's3', 'path' => '*', 'access' => 2]`
&nbsp;&nbsp;&nbsp;&nbsp;`];`
`}`

`if (Auth::user()->role === User::ROLE_ADMIN) {`
` // access to everything under that particular organisation`
&nbsp;&nbsp;&nbsp;&nbsp;`return [`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`['disk' => 's3', 'path' => env('APP_ORG').'*', 'access' => 2],`
&nbsp;&nbsp;&nbsp;&nbsp;`];`
 `}`

`if (Auth::user()->role === User::ROLE_STAFF) {`
` // access to everything under that particular folder ONLY`
&nbsp;&nbsp;&nbsp;&nbsp;`return [`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`['disk' => 's3', 'path' => env('APP_ORG').'/'.$roomPath.'*', 'access' => 2],`
&nbsp;&nbsp;&nbsp;&nbsp;`];`
 `}`

` // if anyone else - 0 access (I needed this part - you might not)`
`return [`
&nbsp;&nbsp;&nbsp;&nbsp;`['disk' => 's3', 'path' => '*', 'access' => 0],`
`];`

Also, in the config/file-manager.php - I use 

`'aclStrategy' => 'whitelist'`

PS: If you're getting aclError when accessing paths that you've already allowed access to - please use something like xDebug to check the paths being compared by the internal functions of this file-manager). I found that my path had an extra "/" and was therefore failing the rules.

## Middleware (Extra Stuff - incase someone comes across a similar situation and needs it)
To make sure the path opened by default, is the same as the url slug - in my case these are rooms IDs.
e.g. - /testapp/<room_id>

I created a middleware (check Laravel documentation to figure out how to create **and register** one properly).
The handle() procedure I used is as follows -

`public function handle($request, Closure $next)    {`
&nbsp;&nbsp;&nbsp;&nbsp;`if (!App::runningInConsole()) {` //to make sure this runs only via the browser
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`$roomPath = // logic to get the folder path name. in my case - it's a regEx based matching on the HTTP_REFERRER`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`Config::set('file-manager.leftPath', env('APP_ORG').'/'.$roomPath);`
&nbsp;&nbsp;&nbsp;&nbsp;`}`
&nbsp;&nbsp;&nbsp;&nbsp;`return $next($request);`
`}`

This middleware was then added to config/file-manager.php as, so that all file-manager routes go through this -

`'middleware' => ['web', 'fm.path'],`

Hope I could help! :)

**Comment by daydevelops** [comment]
Hey @BRAiNCHiLD95 

Just wondering if you had much success with the method you described or if you eventually found a better alternative. I am looking for a way to use this package and manage file/folder access via user permissions.

**Comment by BRAiNCHiLD95** [comment]
> Hey @BRAiNCHiLD95
> 
> Just wondering if you had much success with the method you described or if you eventually found a better alternative. I am looking for a way to use this package and manage file/folder access via user permissions.

Oh it absolutely worked for me! 💯

I had a few specific requirements, and I managed to satisfy them all by tweaking/overriding a few of the top-level package methods as described above, nothing else was needed.

All the best with your implementation, if you find anything that I may have missed (or written poorly)... do add it here for anyone that might come across this in the future.

**Comment by daydevelops** [comment]
Thanks @BRAiNCHiLD95 
In my case, I want to authorize a users access to a particular file based on their permissions, not necessarily role. So I am thinking of doing something like:

```bash
public function getUserID() {
    return Auth::user()->permissions;
}
```

And then for getting the rules array for a user:

```bash
$permissions_access = [
	'files: access proposals' => [
		['disk' => 's3', 'path' => '/proposals/*', 'access' => 3],
		['disk' => 's3', 'path' => '/proposals2/*', 'access' => 3],
	],
	'files: access expenses' => [
		['disk' => 's3', 'path' => '/expenses/*', 'access' => 3],
	],
	...
];

$result = [];

forEach(auth()->user()->permissions as $perm) {
	if (array_key_exists($perm,$permissions_access)) {
		$result = array_merge($result,$permissions_access[$perm]);
	}
}

return $result;
```

One thing I am still confused about is new files/folders that are created by the user. What happens if a user creates a new folder called proposals3? I wont have a rule for that folder. Would the solution be to store the rules in my db then, so I can dd or modify rules whenever the user updates a file/folder? 

I ma going to need to track each file/directory change in my db anyway because I have other data in my db for each file.

**Comment by BRAiNCHiLD95** [comment]
Quick Question before I suggest a possible solution -

Will the newly created folder contain the term "proposals" in it? (i.e. - will the folder names be proposals3, proposals4, proposals5, etc.) and the same with expenses?

If those terms are a constant, your solution might just work without any need to inject new rules.

**Comment by BRAiNCHiLD95** [comment]
> One thing I am still confused about is new files/folders that are created by the user. What happens if a user creates a new folder called proposals3? I wont have a rule for that folder. Would the solution be to store the rules in my db then, so I can dd or modify rules whenever the user updates a file/folder?


Quick Question before I suggest a possible solution -

Will the newly created folder contain the term "proposals" in it? (i.e. - will the folder names be proposals3, proposals4, proposals5, etc.) and the same with expenses?

If those terms are a constant, your solution might just work without any need to inject new rules.

**Comment by daydevelops** [comment]
Thanks for the reply. No, I was just using that as an example. In reality, the user has the ability to create any file/folder and I need a way to assign it a permission. My only idea so far is to prompt the user to do this but I dont like that idea.

**Comment by BRAiNCHiLD95** [comment]
> Thanks for the reply. No, I was just using that as an example. In reality, the user has the ability to create any file/folder and I need a way to assign it a permission. My only idea so far is to prompt the user to do this but I dont like that idea.

Sorry but I have another question -
Could it be possible to have a "top level folder" per section? And everything under it (files, folders) would be accessible to people that have those permissions?

i.e. 

Proposals is the top-level folder for proposals, proposals1, 2, 3 would all be created inside it?

```
$permissions_access = [
	'files: access proposals' => [
		['disk' => 's3', 'path' => 'proposals/*', 'access' => 3],
	],
	'files: access expenses' => [
		['disk' => 's3', 'path' => 'expenses/*', 'access' => 3],
	],
        ..... // etc.
];
```

This approach could be used with another config property that helps "hide" everything that a user doesn't have access to (whitelist?, I think)

**Comment by daydevelops** [comment]
Yes that could work. I could perhaps create all of the top level folders myself and assign permissions to the folders and anything that goes into them.

And then not let my users modify them (create, change, or delete) so that their permissions are constant. 

I am taking some flexibility away from my users, meaning that when they want a new one I will have to step in to create it, but I dont see an easy alternative.
